crates/findex/src/backends/custom/ffi/stores.rs

```diff
@@ -121,21 +138,33 @@ macro_rules! impl_ffi_edx_store {
                     serialized_new_values_len,
                 );
 
-                if err == ErrorCode::Success.code() {
-                    let res = unsafe {
-                        std::slice::from_raw_parts(output_ptr.cast_const(), output_len as usize)
-                            .to_vec()
-                    };
-                    Ok(deserialize_edx_lines(&res)?.into_iter().collect())
-                } else {
-                    Err(Self::Error::Ffi("$name upsert error".to_string(), err))
+                match err {
+                    0 => {
+                        let res = unsafe {
+                            std::slice::from_raw_parts(output_ptr.cast_const(), output_len as usize)
+                                .to_vec()
+                        };
+                        let token_encrypted_value_map :cosmian_findex::TokenToEncryptedValueMap<$value_length>=deserialize_edx_lines(&res)?.into_iter().collect();
+                        tracing::debug!("upsert: exiting with {} values", token_encrypted_value_map.len());
+                        Ok(token_encrypted_value_map)
+                    },
+                    1 => Err(Self::Error::Ffi(format!("'{}' upsert error: buffer too small", $name), err)),
+                    2 => Err(Self::Error::Ffi(format!("'{}' upsert error: missing callback", $name), err)),
+                    3 => Err(Self::Error::Ffi(format!("'{}' upsert error: serialization error", $name), err)),
+                    4 => Err(Self::Error::Ffi(format!("'{}' upsert error: backend error", $name), err)),
+                    _ => Err(Self::Error::Ffi(format!("'{}' upsert error: other error: {err}", $name), err)),
                 }
             }
 
+            #[tracing::instrument(level = "trace", fields(map = %map), ret, err, skip(self))]
             async fn insert(
                 &self,
-                map: HashMap<Self::Token, EncryptedValue<$value_length>>,
+                map: cosmian_findex::TokenToEncryptedValueMap<$value_length>,
             ) -> Result<(), Self::Error> {
+                tracing::debug!("insert: entering: map size: {}", map.len());
+                if map.is_empty() {
+                    return Ok(());
+                }
```

Comment:
My take on this recently is that empty requests are not the nominal case. They may even be considered as application logic error. Adding tests which sole purpose is to speed up this class of requests is therefore rather hurtful imho: it hurts readability and adds a branching.

I know some of these tests are still around, but I am planning to remove them as I see them.

```suggestion
```